aten/src/ATen/BatchedTensorImpl.h

```diff
@@ -0,0 +1,86 @@
+#pragma once
+
+#include <bitset>
+
+#include <ATen/ArrayRef.h>
+#include <ATen/SmallVector.h>
+#include <ATen/Tensor.h>
+
+namespace at {
+
+// We assume this in a few other places in the codebase,
+// but there isn't a centralized definition.
+constexpr int64_t kVmapMaxTensorDims = 64;
+
+struct BatchDim {
+  BatchDim(int64_t dim, int64_t level) : dim_(dim), level_(level) {}
+  int64_t dim() const {
+    return dim_;
+  }
+  int64_t level() const {
+    return level_;
+  }
+ private:
+  int64_t dim_;
+  int64_t level_;
+};
+
+using BatchDims = SmallVector<BatchDim, 5>;
+using BatchDimsRef = ArrayRef<BatchDim>;
+
+struct TORCH_API BatchedTensorImpl : public c10::TensorImpl {
+  explicit BatchedTensorImpl(Tensor value, BatchDims bdims);
+
+  BatchDimsRef bdims() const { return bdims_; }
```

Comment:
does "Ref" part in the name implies that it should return reference?

Reply:
BatchDimsRef is typedef'ed to be an `ArrayRef<BatchDim>`. So yes, it is a reference